app/main.py

```
@@ -1,2 +1,125 @@
+                    self.hash_table[index][1] == hash_key:
+                self.hash_table[index][2] = value
+                break
+            index = (index + 1) % self.capacity
+
+    def resize(self, coefficient):
+        old_table_hash = self.hash_table
+        self.capacity = int(self.capacity * coefficient)
+        if self.capacity < 8:
+            self.capacity = 8
+        self.length = 0
+        self.hash_table = [[] for _ in range(self.capacity)]
+        for element in old_table_hash:
+            if len(element) != 0:
+                self.__setitem__(element[0], element[2])
+
+    def __getitem__(self, key):
+        hash_key = hash(key)
+        index = hash(key) % self.capacity
+        while self.hash_table[index]:
+            if self.hash_table[index][1] == hash_key \
+                    and self.hash_table[index][0] == key:
+                return self.hash_table[index][2]
+            index = (index + 1) % self.capacity
+        raise KeyError(f"{key} not in dictionary")
+
+    def get(self, key, value=None):
+        hash_key = hash(key)
+        index = hash(key) % self.capacity
+        while self.hash_table[index]:
+            if self.hash_table[index][1] == hash_key \
+                    and self.hash_table[index][0] == key:
+                return self.hash_table[index][2]
+            index = (index + 1) % self.capacity
+        return value
+
+    def clear(self):
+        self.__init__()
+
+    def __delitem__(self, key):
+        if 6 < self.length <= self.capacity * (1 / 3):
+            self.resize(1 / 2)
+        hash_key = hash(key)
+        index = hash(key) % self.capacity
+        while self.hash_table[index]:
+            if self.hash_table[index][1] == hash_key \
+                    and self.hash_table[index][0] == key:
+                self.hash_table[index] = []
+                self.length -= 1
+                return self
+            index = (index + 1) % self.capacity
+
```

should raise KeyError if not found a key

---

Yes, but if this method has raise KeyError test is faild. I don't know why. Maybe I have some problem with loop? I use the similar loop in __getitem__

---

you don't need a resize here

---

Deleted resize but it doesn't solve the problem

---

try to write like this:

```
    def __getitem__(self, key):
        hashed_value = hash(key)
        index = hashed_value % self.capacity

        if self.length:
            while self.storage[index]:
                if self.storage[index][0] == key \
                        and self.storage[index][2] is True:
                    raise KeyError

                if hashed_value == self.storage[index][2] and \
                        self.storage[index][0] == key:
                    return self.storage[index][1]
                index = (index + 1) % self.capacity

        raise KeyError
```

---

it's only an example, don't use this code, you should write it for your logic

---

Ok, thanks

---

And this example didn't help me

---

what's the error you got?

---

KeyError